### PR TITLE
Convert tables to markdown

### LINKED ISSUE #5: Table convertion
IMO should be added the capability to convert HTML tables to markdown tables.

### PR DESCRIPTION
**Pull request by tijder**
With this merge request it possible to convert tables to markdown. As inspiration I used https://github.com/mixmark-io/turndown-plugin-gfm/blob/master/src/tables.js. 

This pull request will fix #5 

So this html:
```html
<table>
  <thead>
    <tr>
      <th align="left">Column 1</th>
      <th align="center">Column 2</th>
      <th align="right">Column 3</th>
      <th align="foo">Column 4</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Row 1, Column 1</td>
      <td>Row 1, Column 2</td>
      <td>Row 1, Column 3</td>
      <td>Row 1, Column 4</td>
    </tr>
    <tr>
      <td>Row 2, Column 1</td>
      <td>Row 2, Column 2</td>
      <td>Row 2, Column 3</td>
      <td>Row 2, Column 4</td>
    </tr>
  </tbody>
</table>
```

will be converted to:
```md
| Column 1 | Column 2 | Column 3 | Column 4 |
| :-- | :-: | --: | --- |
| Row 1, Column 1 | Row 1, Column 2 | Row 1, Column 3 | Row 1, Column 4 |
| Row 2, Column 1 | Row 2, Column 2 | Row 2, Column 3 | Row 2, Column 4 |
```

And in github this will showed as:
| Column 1 | Column 2 | Column 3 | Column 4 |
| :-- | :-: | --: | --- |
| Row 1, Column 1 | Row 1, Column 2 | Row 1, Column 3 | Row 1, Column 4 |
| Row 2, Column 1 | Row 2, Column 2 | Row 2, Column 3 | Row 2, Column 4 |

A nice thing of the java `org.jsoup:jsoup` library is that tables without a tbody or thead can also be converted, because jsoup add tbody if none is present. This isn't working/supported by the turndown-plugin-gfm library.

### CODE REVIEW

**Comment by furstenheim** [comment]
Hi,
Thanks @tijder for the pr. I'll need some careful reading on this one. I might take some time to review it, since I have other projects atm.

BTW, already published your previous PR with version 1.1, it's already accessible in maven central

**Comment by tijder** [comment]
Take your time. If you have any questions or comments please let me know. I have added most of the test cases from https://github.com/mixmark-io/turndown-plugin-gfm/blob/master/test/index.html. So you can see what the result of the html to markdown is.

**Comment by tijder** [comment]
@furstenheim  can you please look into this PR?

File: src/main/java/io/github/furstenheim/CopyDown.java [hunk 1]
```diff
@@ -134,7 +131,14 @@ public Rules () {
             addRule("blankReplacement", new Rule((element) -> CopyNode.isBlank(element), (content, element) ->
                     CopyNode.isBlock(element) ? "\n\n" : ""));
             addRule("paragraph", new Rule("p", (content, element) -> {return "\n\n" + content + "\n\n";}));
-            addRule("br", new Rule("br", (content, element) -> {return options.br + "\n";}));
+            addRule("br", new Rule("br", (content, element) -> {
+                // If the br tag is in a table keep the br tag. Otherwise, the table will not work in markdown.
+                if (element.parentNode() != null && Arrays.asList("td", "th").contains(element.parentNode().nodeName().toLowerCase())) {
```
Comment (furstenheim) [inline src/main/java/io/github/furstenheim/CopyDown.java:136]: This looks a bit sketchy. It works at only one level, right? Both up and down

File: src/main/java/io/github/furstenheim/CopyDown.java [hunk 2]
```diff
@@ -336,9 +340,71 @@ public Rules () {
                 }
                 return "![" + alt + "]" + "(" + src + titlePart + ")";
             }));
+            addRule("tableCell", new Rule(new String[]{"th", "td"}, (content, element) -> cell(content, element, false)));
+            addRule("tableRow", new Rule("tr", (content, element) -> {
+                StringBuilder borderCells = new StringBuilder();
+                var alignMap = Map.of("left", ":--", "right", "--:", "center", ":-:");
+
+                if (isHeadingRow(element)) {
+                    for (var i = 0; i < element.childNodes().size(); i++) {
```
Comment (furstenheim) [inline src/main/java/io/github/furstenheim/CopyDown.java:349]: Isn't this brittle? Can't you have an html comment as a child, for example?

File: src/test/resources/tests.json [hunk 3]
```diff
@@ -761,5 +761,65 @@
     },
     "input": "\n<pre><code>\nCode\n\n</code></pre>\n    ",
     "output": "```\n\nCode\n\n```"
+  },
```
Comment (furstenheim) [inline src/test/resources/tests.json:764]: Could you add some tests here where the content inside the table has other htmls, also comments, and things like that?

**Comment by furstenheim** [comment]
I understand that the code is just from the other library, but it looks slightly brittle as it is right now

**Comment by tijder** [comment]
Sorry, I close this pull request for now. Maybe when I have time I will look in to it again.

**Pull request closed without merging**